src/lib.rs

```diff
@@ -942,56 +931,74 @@ impl<'a> PtpCamera<'a> {
             }
-
-            loop {
-                let chunk_size = 256 * 1024;
-                let current_len = self.buf.len();
-                let current_capacity = self.buf.capacity();
-                if current_capacity - current_len < chunk_size {
-                    self.buf.reserve(chunk_size);
-                }
-                let remaining_buf = unsafe {
-                    slice::from_raw_parts_mut(self.buf.get_unchecked_mut(current_len) as *mut _, chunk_size)
-                };
-                let timespec = time::get_time();
-                trace!("Read Data  [{}:{:09}] - length:{:?} remaining:{:?}",
-                       timespec.sec,
-                       timespec.nsec,
-                       current_len,
-                       remaining_buf.len());
-                let len = try!(self.handle.read_bulk(self.ep_in, remaining_buf, timeout));
-                unsafe {
-                    self.buf.set_len(current_len + len);
-                }
-                // debug!("new buf len [{:?}] into {:?}", self.buf.len(), remaining_buf.len());
-                if len < remaining_buf.len() {
-                    break;
-                }
+            match container.kind {
+                PtpContainerType::Data => {
+                    data_phase_payload = payload;
+                },
+                PtpContainerType::Response => {
+                    if container.code != StandardResponseCode::Ok {
+                        return Err(Error::Response(container.code));
+                    }
+                    return Ok(data_phase_payload);
+                },
+                _ => {}
             }
+        }
+    }
 
-            let (mtype, mut msg) = try!(PtpTransaction::parse(&self.buf));
+    // helper for command() above, retrieve container info and payload for the current phase
+    fn read_txn_phase(&mut self) -> Result<(PtpContainerInfo, Vec<u8>), Error> {
+        // buf is stack allocated and intended to be large enough to accomodate most
+        // cmd/ctrl data (ie, not media) without allocating. payload handling below
+        // deals with larger media responses.
+        let n;
+        let mut buf = [0u8; 8 * 1024];
+        let timeout = Duration::from_secs(2);
 
-            if mtype == PtpContainerType::Data && msg.is_response(&transaction) {
-                data = Some(msg.data);
-            } else if mtype == PtpContainerType::Response && msg.is_response(&transaction) {
-                if let Some(data) = data {
-                    msg.data = data;
-                }
-                
-                if msg.code != StandardResponseCode::Ok {
-                    return Err(Error::Response(msg.code));
-                }
-                
-                return Ok(msg.data);
+        unsafe {
+            let bufslice = slice::from_raw_parts_mut(&mut buf[0] as &mut u8, buf.len());
```

Comment:
You can just pass `&mut buf[..]` to read_bulk and don't need `from_raw_parts_mut`. You also don't need unsafe{} because you're no longer using uninitialized memory. If you want to avoid zero-initializing buf, you can initialize it with `std::mem::uninitialized()`.